tests/test_target_database.py

```diff
@@ -1,46 +1,51 @@
-        target_data = target_database.get_target_data()
-        self.assertIsNone(target_data, "Target data is expected to be empty on an error")
+    def test_401(self, mock_request):
+        """Given a 401 error code, TargetAPIError is raised."""
+        mock_request.get(
+            target_database._TARGET_API, status_code=401
+        )
+        with self.assertRaises(target_database.TargetAPIError):
+            target_database.get_target_data()
 
     @requests_mock.mock()
-    def test_401(self, mock_request):
-        """Given a 401 response, target data is not set."""
-        mock_request.get(target_database._TARGET_API, status_code=401)
-        target_data = target_database.get_target_data()
-        self.assertIsNone(target_data, "Target data is expected to be empty on an error")
+    def test_404(self, mock_request):
+        """Given a 404 error code, TargetAPIError is raised."""
+        mock_request.get(
+            target_database._TARGET_API, status_code=404, text="Not Found"
+        )
+        with self.assertRaises(target_database.TargetAPIError):
+            target_database.get_target_data()
 
     @requests_mock.mock()
     def test_200_invalid_json(self, mock_request):
-        """Given a valid response but invalid json, target data is not set."""
-        mock_request.get(target_database._TARGET_API, text="invalid_json")
-        target_data = target_database.get_target_data()
-        self.assertIsNone(target_data, "Target data is expected to be empty on an error")
+        """Given a valid response but invalid json, JSONDecodeError is raised."""
+        mock_request.get(target_database._TARGET_API, json=None)
+        with self.assertRaises(target_database.ResponseJSONError):
+            target_database.get_target_data()
 
     @requests_mock.mock()
-    def test_200_no_data(self, mock_request):
-        """Given a valid response but no json data, target data is not set."""
-        mock_request.get(target_database._TARGET_API, json={})
-        target_data = target_database.get_target_data()
-        self.assertIsNone(target_data, "Target data is expected to be empty on an error")
+    def test_200_no_data_field(self, mock_request):
+        """Given a valid response but no data field, ResponseJSONError is raised."""
+        mock_request.get(target_database._TARGET_API, json={"notdata": []})
+        with self.assertRaises(target_database.ResponseJSONError):
+            target_database.get_target_data()
 
     @requests_mock.mock()
     def test_200_value_data(self, mock_request):
         """Given a valid response, target data is set from the returned json."""
         mock_request.get(target_database._TARGET_API, json={"data": 42})
         target_data = target_database.get_target_data()
-        self.assertEqual(42, target_data, "Target data is should match the contents of the target API data")
+        self.assertEqual(target_data, 42, "Target data should match the contents of the target API data")
```

Comment:
The form of the assert methods are `(expected, actual, explanation)`
```suggestion
        self.assertEqual(42, target_data, "Target data should match the contents of the target API data")
```

Reply:
This is wrong and never explicitly stated in python docs.

1. `self.assertEqual(actual, expected)` reads better - "I expect server response to be this dict" vs "I expect this dict to be server response".
2. It is consistent with how you do conditionals, hence not an exception to memorize: `if foo == 2`, not `if 2 == foo`.
3. Observe:
```
    def test_dicts(self):
        # Result of calling subject under test
        subject = {
            'a': True,
            'b': False,
            'c': None,
        }
        self.assertEqual(
            subject, {
>               'd': set(),
            }
        )
E       AssertionError: {'a': True, 'b': False, 'c': None} != {'d': set()}
E       - {'a': True, 'b': False, 'c': None}
E       + {'d': set()}
```
The diff is true to the fact that lines with `-` are missing from my subject and lines with `+` are unexpected additions. Other way round it's just weird.

Reply:
Yes, I agree - in the python docs it reads `(first, second)` and some linters advise to put the actual & expected the other way, too. Honestly didn't have the energy to push back on this as I thought it would spawn another 50 review comments

Reply:
@muchzill4 @rwalton-arm I agree that the documentation is far from clear but I base my request for change on the apparent implementation of `assertEqual`.

Changing the code to:

`self.assertEqual(41, target_data, "Target data should match the contents of the target API data")`

Gives the following warning:

```
Target data should match the contents of the target API data
42 != 41

Expected :41
Actual   :42
```

Reply:
@muchzill4 Looks like the conclusion from your example in point 3. above is the wrong way round :-). If 'd' is an unexpected addition then, by definition, 'subject' must be the thing that you are expecting to see. Hence the expected thing goes first in the call, and the actual thing goes second. This is the same as in jUnit, but differs from cUnit. Aren't standards wonderful :-)

Reply:
@madchutney did you add `Expected` and `Actual`? I'm not able to make `pytest` nor `unittest` yield such output.

@mark-edgeworth I don't follow. :) Anyway, the more I look at 3., the more I realize it is fragile as it could be argued both ways, so you're right in that it's a moot point.

Reply:
FWIW I had a browse in the `cpython` and the tests for `unittest` itself use `actual, expected`. 😈

Reply:
@muchzill4 I just ran them in PyCharm. Personally I don't like the ordering, but 🤷‍♂ it is what it is.